### ISSUE TITLE
More control on composition

### ISSUE THREAD

**Issue by koresar** [stampit-org/stamp-specification#21]
I wish there was a feature when I, as a stamp author, can get access to both **left**, **right** and **resulting** stamps in a single context.

Fanasy code:
```js
const CompositionControl = compose({
  postCompose({left, right, result}) {
    if (left.methods.foo && right.methods.foo) { // Dummy implementation, don't pay attention please
      result.methods._foo1 = left.methods.foo;
      result.methods._foo2 = right.methods.foo;
      result.methods.foo = function(...args) { this._foo1(...args); this._foo2(...args); }
    }
  }
});

const FooStamp = CompositionControl.compose({ methods: {
  foo() { console.log('foo'); }
}});

const BarStamp = compose({ methods: {
  foo() { console.log('OVERRIDE'); }
}});

const Final = compose(FooStamp, BarStamp);

Final(); // will print both "foo" and "OVERRIDE"
```

This could be beneficial to handle/maintain/log collision issues.

Thoughts?

**Comment by JosephClay** [comment]
It appears that this would require the implementor to compose in the correct order.

e.g. this may work (using your above example):
```js
compose(FooStamp, BarStamp, BazStamp);
```
but this may not:
```js
compose(BarStamp, BazStamp, FooStamp);
```

If so, then the method is leaking implementation logic and would be extremely difficult to debug.

**Comment by koresar** [comment]
Please, do not look for flaws. Seek opportunity.

This wil give us the following:
1) Collision control (many people dream about for few months already).
2) "Debug mode" for bug search.
3) List of all stamps participated in an object instance creation.
4) ... etc etc etc


P.S.
How about that?
```js
CompositionControl.compose(BarStamp, FooStamp);
CompositionControl.compose(FooStamp, BarStamp);
```

**Comment by ericelliott** [comment]
Is there a reason that user's can't handle this stuff in initializers? Initializers get passed the stamp and the stamp arguments, which in my mind gives you everything you need to override the default composition behavior in any way you want without modifying the composable API... right?

**Comment by ericelliott** [comment]
I'm not convinced this is needed at all. See, for instance:

```js
import stampit from 'stampit'
import Sight from './sight'
import Sound from './sound'

const Senses = stampit().init(({ instance }) => {
  const sight = Sight();
  const sound = Sound();

  instance.manifest = () => {
    sight.manifest();
    sound.manifest();
  }

  return instance;
});
```

Initializers give you the power to compose multiple stamps in any way you want using the stamp API. If you need to do things that aren't supported by the default compose logic, you should do so in a stamp, not by adding complexity to the compose API.

We should show examples of how to do things like that using stamps in advanced examples.

**Comment by ootoovak** [comment]
Maybe there is an opportunity here for some tiny Stampit modules instead of just examples in files in this repo? Could have links to all the NPM modules (maybe even namespaced?) that provide more specialized functionality. That way Stampit remains small and generalised and you start an ecosystem of Stamps. So in the example @ericelliott put above it could be a Stamp module that gives a new `.hasARelationship(Relationships)` static method or something.

```javascript
import stampit from 'stampit'
import Sight from './sight'
import Sound from './sound'

const Senses = stampit().hasARelationship(Sight, Sound)
  .init(({ instance }) => {
    instance.manifest = () => {
      instance.relationships.forEach((relationship) => {
        relationship.maifest()
      })
    }
    return instance
  })
```

**Comment by koresar** [comment]
@ericelliott Take a look at the endless opportinities this feature would open. Stop looking for the workaround over the example on the top. 

Okay, I asked not to pay attention to the code, but now let's talk about code only.

Please, rewrite the following using the `init()`.

1) Printing **all** overlapping methods.
```js
const OverlapMethodsPrint = compose({
  postCompose({left, right, result}) {
    const ovelapped = _.intersection(_.keys(left.methods), _.keys(right.methods));
    if (overlapped.length) {
      console.log(overlapped);
    }
  }
});
const abc1 = compose({methods: { a(){}, b(){}, c(){} }});
const abc2 = compose({methods: { a(){}, b(){}, c(){} }}); 
OverlapMethodsPrint.compose(abc1, abc2); // prints "[ 'a', 'b', 'c' ]"
```

2) Throwing error when statics overlap (see https://github.com/stampit-org/stampit/issues/148 )
```js
const OverlapStaticsPrint = compose({
  postCompose({left, right, result}) {
    const ovelapped = _.intersection(_.keys(left.statics), _.keys(right.statics));
    if (overlapped.length) {
      throw new Error('The statics overlap: ', overlapped.join());
    }
  }
});
const EE = convertConstructor(EventEmitter);
const HaveStaticInit = compose({static: {init(){ console.log('INITIALIZED'); }}});
OverlapStaticsPrint.compose(EE, HaveStaticInit); // expect throw Error
```

3) Print all the stamps participated in instance creation as a TREE:
```js
const CollectOriginStamps = composE({
  postCompose({left, right, result}) {
    result._stamps = _.unique(left._stamps, right._stamps, [left, right]);
    left._level = left._level || 0;
    right._level = right._level || 0;
    result._left = left;
    result._right = right;
    result._level = _.max(left._level, right._level) + 1;
  }
});

const S1 = compose();
const S2 = compose();
const S3 = compose();
const TreePrinter = ({ 
  init({stamp}) {
    console.log(`Total ${stamp._stamps.length} stamps`);
    console.log('Stamps dependency TREE');
    print = (stamp) => {
      if (!stamp) { return; }
      console.log(stamp._left._level || '', stamp._right._level || '');
      print(stamp._left);
      print(stamp._right);
    }
    print(stamp);
  }
});

compose(S1, S2, S3, TreePrinter)();
```

There is no way to "do so in a stamp". The API does not give us both parent stamps in the `init()` function.

So, I wonder, can anyone else see the unlimited potential for the andvaced scenarios like this? Debugging could be simply perfect!

**Comment by ootoovak** [comment]
Hay @koresar not sure if the above was in response to my post but if it was yeah I can see the potential issues, I am happy using the solutions suggested here and in Gitter. I guess I just thinking aloud after reading about the `.static` API and how that might work in a module ecosystem. Appreciate the feedback and the work you have done on the project thus far. :+1:

**Comment by koresar** [comment]
@ootoovak thanks mate. :)

**Comment by JosephClay** [comment]
> There is no way to "do so in a stamp"

1 can be done entirely outside of a `postCompose` method and be much more clear. 2 is literally an error thrower for `Object.assign`. 3 is 

This seems to all be implementation detail that should be handled by the implementor, not a stamp. 

> The API does not give us both parent stamps in the init() function.

I don't think it should. This encourages breaking core concepts of stamps. Having access to the parent to make calls is as bad as `super()` or `Foo.prototype.call(this, arguments)`.

If debug-ability is the goal, then lets aim for a way to do that safely. Post stamp mutation is going to be worse for debugging, not better.

> ...could be a Stamp module that gives a new `.hasARelationship(Relationships)` static method or something

I could see this method as being useful almost as an `instanceof` method.

```js
import stampit from 'stampit';
import Sight from './sight';
import Senses from './senses';

const sense = Senses();

stampit.hasARelationship(sense, Sight) // true

import Hand from './hand';
stampit.hasARelationship(sense, Hand) // false
```

**Comment by ericelliott** [comment]
## Complicating an Interface

Any time you want to add features to an interface, you need to justify those features, because:

1. More API surface area means more room for bugs.
2. More room for bugs means more room for security hazards. In a lib, multiply this hazard by all apps using it.
3. More API surface area means more to learn. Sharper learning curve.
4. Sharper learning curve = more intimidating to new users.

In other words, new features have both negative and (sometimes) positive impacts on that API.

So the challenge isn't, "show me how to do this with the current API."

The challenge is "show me valid use cases compelling enough to justify the costs listed above.

We should be looking for ways to *simplify* the stamp API, rather than ways to make it more complex.

**Comment by koresar** [comment]
@JosephClay @ericelliott Agree to both of you. API should be short and simple. 
Thank you very much for the detailed and thorough examplainations. Sorry I took your time.

Closing this issue now.

P.S. At the end of the day, if someone would need `postCompose` or alike he/she can add it in they own composables implementation. Similar to Bluebird. It is PromisesA+ compliant but also has many more useful functions (like `.spread`, etc).

**Comment by ericelliott** [comment]
:+1: 

I've been using stampit longer than anybody, and I've never felt a need for it. In the few use-cases where it might have came in handy, I was able to get by with a solution like [the one above](https://github.com/stampit-org/stamp-specification/issues/21#issuecomment-133916305). =)
